### PR TITLE
token-auth: add vault support

### LINKED ISSUE #1537: Security: Make it possible to store API-Token-Key in Vault (instead of configuration)
# Feature Request

At the time of writing the API-Token, for example used to authenticate at the Data Management API, is part of the connector configuration. It would be more secure if this information is moved into the Vault and the configuration only contains an alias.

## Which Areas Would Be Affected?
- `Auth-Tokenbased` Extension

## Why Is the Feature Desired?
- Improve security

## Solution Proposal
- Move auth key from settings to Vault
or
- Add configuration to store auth key in vault, so that it's possible to have the key in settings **or** Vault

## Type of Issue
- enhancement

## Checklist

- [x] assigned appropriate label?
- [x] **Do NOT select a milestone or an assignee!**

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

### PR DESCRIPTION
**Pull request by DominikPinsel**
## What this PR changes/adds

Extension for token based authentication can now retrieve the authentication key from a vault.

see issue https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1537

## Why it does that

- described in issue

## Further notes

- 

## Linked Issue(s)

Closes #1537 

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

### CODE REVIEW

File: extensions/api/auth-tokenbased/src/main/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtension.java [hunk 1]
```diff
@@ -9,11 +9,16 @@
  *
  *  Contributors:
  *       Microsoft Corporation - initial API and implementation
+ *       Mercedes-Benz Tech Innovation GmbH - add README.md
+ *       Mercedes-Benz Tech Innovation GmbH - authentication key can be retrieved from vault
```
Comment (florianrusch-zf) [inline extensions/api/auth-tokenbased/src/main/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtension.java:13]: Can we merge this? Or is the first one really necessary?

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

**Comment by florianrusch-zf** [comment]
Ah sorry, didn't saw, that the PR is still a draft 🙈

**Comment by DominikPinsel** [comment]
> 



> Ah sorry, didn't saw, that the PR is still a draft see_no_evil

Feedback is always welcome :)

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    private static final String REFLECTION_VAULT_FIELD_NAME = "vault";
+
+    private static final String AUTH_SETTING_APIKEY = "edc.api.auth.key";
+
+    private static final String AUTH_SETTING_APIKEY_ALIAS = "edc.api.auth.key.alias";
+
+    private static final String VAULT_KEY = "foo";
+
+    // mocks
+    private Vault vaultMock = null;
```
Comment (ndr-brt) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:36]: The mocks could be initiated directly here, the `null` initialization is unnecessary and a bit misleading.

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    private TokenBasedAuthenticationExtension extension;
+    private static final String REFLECTION_VAULT_FIELD_NAME = "vault";
+
+    private static final String AUTH_SETTING_APIKEY = "edc.api.auth.key";
+
+    private static final String AUTH_SETTING_APIKEY_ALIAS = "edc.api.auth.key.alias";
+
+    private static final String VAULT_KEY = "foo";
+
+    // mocks
```
Comment (ndr-brt) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:35]: unnecessary comment

**Review by ndr-brt — APPROVED** [review]

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java (hunk 3 above)
Comment (DominikPinsel) [reply extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:35]: done :)

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java (hunk 2 above)
Comment (DominikPinsel) [reply extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:36]: done :)

File: extensions/api/auth-tokenbased/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Token Based Authentication Service
+
+The token based authentication service extension is used to secure connector APIs. These APIs are not protected by the `AuthenticationService` by default. To find out how a specific API is protected please consolidate its documentation.
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/README.md:3]:
```suggestion
The token based authentication service extension is used to secure connector APIs. These APIs are not protected by the `AuthenticationService` by default. To find out how a specific API is protected please consult its documentation.
```

File: extensions/api/auth-tokenbased/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+```
+
+## Configuration
+
+| Key | Description | Required |
+|:---|:---|:---|
+| edc.api.auth.key | API Key Header Value | false |
+| edc.api.auth.key.alias | Secret name of the API Key Header Value, stored in the vault | false |
+
+If no authentication key is defined, a random value is generated and printed out into the logs.
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/README.md:19]: I think it should be made clear
- that the API key can be stored in the `Vault` _and_ in the configuration
- what the order of precedence is, i.e. what happens if there's a key in the vault AND the config, etc.

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    private TokenBasedAuthenticationExtension extension;
+    private static final String REFLECTION_VAULT_FIELD_NAME = "vault";
+
+    private static final String AUTH_SETTING_APIKEY = "edc.api.auth.key";
+
+    private static final String AUTH_SETTING_APIKEY_ALIAS = "edc.api.auth.key.alias";
+
+    private static final String VAULT_KEY = "foo";
+
+    private Vault vaultMock = Mockito.mock(Vault.class);
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:35]: nit: static import
```suggestion
    private Vault vaultMock = mock(Vault.class);
```

Please do this everywhere in the file.

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        extension = new TokenBasedAuthenticationExtension();
+
+        var monitorMock = Mockito.mock(Monitor.class);
+
+        Mockito.when(serviceExtensionContextMock.getMonitor()).thenReturn(monitorMock);
+        Mockito.when(vaultMock.resolveSecret(VAULT_KEY))
+                .thenReturn("foo");
+
+        // "inject" vault
+        var field = TokenBasedAuthenticationExtension.class.getDeclaredField(REFLECTION_VAULT_FIELD_NAME);
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:50]: Please use `DependencyInjectionExtension` to do the injecting.

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        field.setAccessible(true);
+        field.set(extension, vaultMock);
+    }
+
+    @Test
+    public void testPrimaryMethod_LoadKeyFromVault() {
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:56]:
```suggestion
    public void testPrimaryMethod_loadKeyFromVault() {
```

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                .getSetting(Mockito.eq(AUTH_SETTING_APIKEY), Mockito.anyString());
+
+        Mockito.verify(serviceExtensionContextMock, Mockito.times(1))
+                .getSetting(AUTH_SETTING_APIKEY_ALIAS, null);
+
+        Mockito.verify(vaultMock, Mockito.times(1)).resolveSecret(VAULT_KEY);
+    }
+
+    @Test
+    public void testSecondaryMethod_LoadKeyFromConfig() {
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:72]:
```suggestion
    public void testSecondaryMethod_loadKeyFromConfig() {
```

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+        extension.initialize(serviceExtensionContextMock);
+
+        Mockito.verify(serviceExtensionContextMock, Mockito.times(1))
+                .getSetting(Mockito.eq(AUTH_SETTING_APIKEY), Mockito.anyString());
+
+        Mockito.verify(serviceExtensionContextMock, Mockito.times(1))
+                .getSetting(AUTH_SETTING_APIKEY_ALIAS, null);
+
+        Mockito.verify(vaultMock, Mockito.times(0)).resolveSecret(Mockito.anyString());
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:85]: better to use:
```suggestion
        Mockito.verify(vaultMock, never()).resolveSecret(Mockito.anyString());
```

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
Couple of minor nits apart from the dependency injection stuff

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+        extension.initialize(serviceExtensionContextMock);
+
+        verify(serviceExtensionContextMock, times(1))
+                .getSetting(eq(AUTH_SETTING_APIKEY), anyString());
+
+        verify(serviceExtensionContextMock, times(1))
+                .getSetting(AUTH_SETTING_APIKEY_ALIAS, null);
+
+        verify(vaultMock, times(0)).resolveSecret(anyString());
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:93]:
```suggestion
        verify(vaultMock, never()).resolveSecret(anyString());
```

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+        extension.initialize(serviceExtensionContextMock);
+
+        verify(serviceExtensionContextMock, never())
+                .getSetting(eq(AUTH_SETTING_APIKEY), anyString());
+
+        verify(serviceExtensionContextMock, times(1))
+                .getSetting(AUTH_SETTING_APIKEY_ALIAS, null);
+
+        verify(vaultMock, times(1)).resolveSecret(VAULT_KEY);
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:76]: not needed, as `times(1)` is the default
```suggestion
        verify(vaultMock).resolveSecret(VAULT_KEY);
```

File: extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    public void testPrimaryMethod_loadKeyFromVault() {
+        setAuthSettingApiKeyAlias(VAULT_KEY);
+        setAuthSettingApiKey("bar");
+
+        extension.initialize(serviceExtensionContextMock);
+
+        verify(serviceExtensionContextMock, never())
+                .getSetting(eq(AUTH_SETTING_APIKEY), anyString());
+
+        verify(serviceExtensionContextMock, times(1))
```
Comment (paullatzelsperger) [inline extensions/api/auth-tokenbased/src/test/java/org/eclipse/dataspaceconnector/api/auth/TokenBasedAuthenticationExtensionTest.java:73]:
```suggestion
        verify(serviceExtensionContextMock)
```

**Review by paullatzelsperger — APPROVED** [review]

**Pull request merged**